### DISCUSSION TITLE
Framework agnostic Python-Library for Unpoly

### DISCUSSION THREAD

**Discussion by apollo13** [unpoly/unpoly#262] [Show and tell]
Over the last day I managed to fully (I think) copy unpoly-rails and put it into a python library: https://gitlab.com/rocketduck/python-unpoly

Test coverage is close to 100% and I'll be working on docs over the next few days; but the handling is relatively easy (for Django it ships a middleware which basically what's shown below already):
```py
from unpoly import Unpoly
from unpoly.contrib.django import DjangoAdapter

# do this per request (for instance in a middleware)
adapter = DjangoAdapter(request)
request.up = Unpoly(adapter)
# view logic coming here
response = some_view(request)
# at the end of the middleware write out headers etc
up.finalize_response(response)
```

The `Unpoly` class exposed access to the context, cache etc stuff from the protocol. The library is purposefully minimal and higher level abstractions could be built on it. For instance `django-unpoly` by @jwaschkau (:wave: hi ;)) could use it for the low level plumbing and focus itself solely on high-level interfaces for Django itself. 

The library is framework agnostic and can be used in any framework via a simple adapter; I'd love to see people add support for Flask and others. The existing adapter for Django looks like this:
```py
class DjangoAdapter(BaseAdapter):
    def __init__(self, request):
        self.request = request

    def request_headers(self):
        return self.request.headers

    def request_params(self):
        return self.request.GET

    def redirect_uri(self, response):
        return (
            response.headers.get("Location")
            if 300 <= response.status_code < 400
            else False
        )

    def set_redirect_uri(self, response, uri):
        response.headers["Location"] = uri

    def set_headers(self, response, headers):
        for k, v in headers.items():
            response.headers[k] = v

    def set_cookie(self, response, needs_cookie=False):
        if needs_cookie:
            response.set_cookie(UP_METHOD_COOKIE, self.method)
        elif UP_METHOD_COOKIE in self.request.COOKIES:
            response.delete_cookie(UP_METHOD_COOKIE)

    @property
    def method(self):
        return self.request.method

    @property
    def location(self):
        return self.request.get_full_path_info()
```

**Comment by adam12** [comment]
Nice job! I'm trying to find time to wrap up my plain Ruby adapter and it's proven to be challenging :P
